runtime/bundles/org.eclipse.e4.core.di/src/org/eclipse/e4/core/internal/di/InjectorImpl.java

```diff
@@ -148,38 +149,32 @@ private void internalInject(Object object, PrimaryObjectSupplier objectSupplier,
 
 	private void rememberInjectedObject(Object object, PrimaryObjectSupplier objectSupplier) {
 		synchronized (injectedObjects) {
-			List<WeakReference<?>> list = injectedObjects.computeIfAbsent(objectSupplier, k -> new ArrayList<>());
-			for (WeakReference<?> ref : list) {
-				if (object == ref.get())
-					return; // we already have it
+			Set<IdentityWeakReference<? extends Object>> set = injectedObjects.computeIfAbsent(objectSupplier,
+					k -> new LinkedHashSet<>());
+			IdentityWeakReference<?> ref = new IdentityWeakReference<>(object);
+			if (!set.contains(ref)) {
```

Comment:
Sorry I keep noticing things...  

There is no need to test contains before doing an add; a set won't add a duplicate.  So probably this whole thing can be:
```
injectedObjects.computeIfAbsent(objectSupplier, k -> new LinkedHashSet<>()). add(new IdentityWeakReference<>(object);
```
Then we don't need a name for the "set".

Reply:
I also noticed this, I was thinking about if the check is faster than the adding, I did not really find something about this, so i will change to the one liner

Reply:
In general the one liner is faster.

- **Case 1**: Item is not in the set
- **Case 2**: Item is in the set

With "if/then":

1. This will result in two checks if the item is present (one from the if and one from the add) and one add operation to the internal datastructure
2. This will result in one checks if the item is present (one from the if)

With "oneliner":

1. One check and one add to the internal datastructure
2. one check (internally) and early exit

so for case (1) the oneliner saves one additional check while for (2) it is equally.

Reply:
Yes, you are right. I was wondering if the contains check from the set or contains is faster, because those seam to be different implementations. But maybe I thought too much about it.